### PR TITLE
K8s: bug fixes for file storage to allow for watcher initialization on startup

### PR DESCRIPTION
**Pull request by charandas**
**What is this feature?**

Fixes a range of configuration and other bugs for file storage.

The RESTOptionsGetter, optionally, allows one to specify initial directories to bootstrap which read/watch operations on start aren't able to establish/mkdir and otherwise cause failures in aggregation init.

Other tidbits in this changeset:

* Guaranteed Update was missing an ensureDir.
* Refactored the variable names to make more sense. Paths are paths, not just names.
* The DirPath helper on storage is quite different from standard library `filepath.Dir`. Both have their uses, so left a comment noting the difference.

**Why do we need this feature?**

To use with externalnames and apiregistration until unified is ready.

**Who is this feature for?**

Grafana App Platform

**Which issue(s) does this PR fix?**:

Fixes #

**Special notes for your reviewer:**

Please check that:
- [ ] It works as expected from a user's perspective.
- [ ] If this is a pre-GA feature, it is behind a feature toggle.
- [ ] The docs are updated, and if this is a [notable improvement](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/#how-to-determine-if-content-belongs-in-a-whats-new-document), it's added to our [What's New](https://grafana.com/docs/writers-toolkit/writing-guide/contribute-release-notes/) doc.

### CODE REVIEW

**Review by toddtreece — APPROVED** [review]
🤦

File: pkg/apiserver/storage/file/file.go [hunk 1]
```diff
@@ -306,6 +307,13 @@ func (s *Storage) Watch(ctx context.Context, key string, opts storage.ListOption
 // match 'opts.ResourceVersion' according 'opts.ResourceVersionMatch'.
 func (s *Storage) Get(ctx context.Context, key string, opts storage.GetOptions, objPtr runtime.Object) error {
 	filename := s.filePath(key)
+
+	// Since it's a get, check if the dir exists and return early as needed
+	dirname := filepath.Dir(filename)
+	if _, err := os.Stat(dirname); err != nil {
```
Comment (toddtreece) [inline pkg/apiserver/storage/file/file.go:313]: thoughts on making a util function for this?

something like:

```go
func exists(path string) (bool, error) {
    _, err := os.Stat(path)
    if err == nil {
        return true, nil
    } else if os.IsNotExist(err) {
        return false, nil
    } else {
        return false, err
    }
}
```

File: pkg/apiserver/storage/file/file.go [hunk 2]
```diff
@@ -306,6 +307,13 @@ func (s *Storage) Watch(ctx context.Context, key string, opts storage.ListOption
 // match 'opts.ResourceVersion' according 'opts.ResourceVersionMatch'.
 func (s *Storage) Get(ctx context.Context, key string, opts storage.GetOptions, objPtr runtime.Object) error {
 	filename := s.filePath(key)
+
+	// Since it's a get, check if the dir exists and return early as needed
+	dirname := filepath.Dir(filename)
```
Comment (toddtreece) [inline pkg/apiserver/storage/file/file.go:312]: should this be `dirname := s.dirPath(key)` just to keep things consistent with list? it seems like little differences can make debugging confusing if one or the other stops working

Comment (charandas) [reply pkg/apiserver/storage/file/file.go:311]: Thanks! Yeah we don't want bugs on different OSes. I now have a windows VM if we ever need to test something over there.

**Pull request merged**